pollfd fix, round 2

vtjnash opened the pull request:
this is a rewrite of the poll_fd, poll_file, and watch_file APIs to more directly expose the libuv API and eliminate the callback legacy interface. since `FDWatcher` objects are now refcounted, this also solves #4401.

this replaces https://github.com/JuliaLang/julia/pull/7275

ping: @Keno

amitmurthy committed:
revised poll_fd tests

amitmurthy committed:
standardized event property testing

vtjnash committed:
fix poll_fd implementation (close #4401)

this eliminates the callback interface to the filesystem poll functions,
to be consistent with the changes to the rest of base

additionally, this exposes the full libuv API for watch_file and
poll_file, rather than returning inconsistent return values from these two functions

this fixes FDWatcher to ensure that it is unique for any given fd,
(to avoid a libuv assertion). it must be manually closed before closing
fd, or the applications behavior is undefined.

vtjnash commented:
cross-ref: this will resolve some https://github.com/JuliaLang/julia/issues/10293

vtjnash commented:
this closes https://github.com/JuliaLang/julia/issues/7840 (when merged)

vtjnash commented on base/stream.jl:
```diff
@@ -595,6 +597,7 @@ function _uv_hook_close(t::Timer)
     unpreserve_handle(t)
     disassociate_julia_struct(t)
     t.handle = C_NULL
+    notify(t.cond)
```
this fix is the cause of the warning in the `repl.jl` test: the Timer callback on line 100 of test/repl.jl is supposed to accept an argument `t`

JeffBezanson commented on base/poll.jl:
```diff
@@ -56,293 +26,381 @@ const FD_TIMEDOUT = 4
-    cb::Callback
-    function PollingFileWatcher(cb, file)
+    active::Bool
+    function PollingFileWatcher(file::AbstractString, interval::Float64=5.007) # same default as nodejs
```
What a weird constant!

vtjnash replied:
i don't know where it comes from either (https://nodejs.org/api/fs.html#fs_fs_watchfile_filename_options_listener)

JeffBezanson commented on base/poll.jl:
```diff
@@ -56,293 +26,381 @@ const FD_TIMEDOUT = 4
+    active::Bool
+    function PollingFileWatcher(file::AbstractString, interval::Float64=5.007) # same default as nodejs
+        if interval < 2
+            warn("For maximum portability, use multi-second intervals with PollingFileWatcher")
```
Could we either make this an error, or give an accurate error based on what the current platform actually supports?

vtjnash replied:
i just copied the warning from the libuv docs (http://docs.libuv.org/en/latest/fs_poll.html#c.uv_fs_poll_start)

JeffBezanson replied:
Humph, typical wishy-washy nonsense. Also, it says to use multi-second intervals, but that sub-second intervals are the problematic ones. Ambiguous for values between 1 and 2. I think we should just give an error for values less than 1.

vtjnash replied:
the libuv implementation is to call `stat` in a loop off a uv_timer_t and invoke the callback if any of the fields are not identical. other than being extremely inefficient for small times (e.g, from what I could tell, osx was only giving me 1sec resolution for mtime), I actually don't see an issue.

JeffBezanson replied:
I just prefer not to annoy people with warnings. Realistically, somebody is not going to happily run a program that scrolls that warning repeatedly, so it might as well be an error. Especially since the warning depends directly on a passed argument, we need to decide whether the argument value is allowed or not.

JeffBezanson commented on doc/stdlib/io-network.rst:
```diff
@@ -723,22 +723,33 @@ Network I/O
+
+   Monitor a file for changes by polling every ``interval_seconds`` seconds until a change occurs or ``timeout_s`` seconds have elapsed.
+
+   Returns a pair of ``StatStruct`` objects when a change is detected. (using ``watch_file`` is preferred, if supported)
```
This has a very different return value from `watch_file`, so you can't just switch from one to the other. At the same time, if `watch_file` is better, why does `poll_file` exist? Can we pick for you? If not, the docs should describe the tradeoffs so you can pick.

vtjnash replied:
this was also copied from nodejs: https://nodejs.org/api/fs.html#fs_caveats (although i guess there, they are nice enough to tell you to do a comparison of `curr.mtime and prev.mtime` to make this compatible)

JeffBezanson commented on doc/stdlib/io-network.rst:
```diff
@@ -723,22 +723,33 @@ Network I/O
    ``timedout``, giving the result of the polling.
 
-.. function:: poll_file(s, interval_seconds::Real, seconds::Real)
+   Internally, this uses a ``FDWatcher``.
```
Do you need to know about this type? Is there anything useful you can do with it?

vtjnash replied:
you could create these types manually and iterate a wait loop somewhat more efficiently (at the risk of losing events, since they are edge-based)

JeffBezanson replied:
Well, we need to make a call here. If you feel using these types directly is not a great idea, we should just leave them as an implementation detail.

vtjnash replied:
it's probably best to just consider them an implementation detail for now then. it's easier to revist that later than the reverse.

JeffBezanson commented:
Awesome, this is a relief!

I believe the only piece of #10293 left is SingleAsyncWork?

tkelman commented on base/libc.jl:
```diff
@@ -23,6 +23,15 @@ dup(src::RawFD,target::RawFD) = systemerror("dup",-1==
+end
+
+@unix_only _get_osfhandle(fd::RawFD) = fd
+@windows_only _get_osfhandle(fd::RawFD) = WindowsRawSocket(ccall(:_get_osfhandle,Ptr{Void},(Cint,),fd.fd))
```
this'll need to be qualified with `Libc.` now anywhere it's used, like `base/stream.jl` and `base/mmap.jl`

tkelman committed:
Qualify Libc._getosfhandle

tkelman commented on test/pollfd.jl:
```diff
@@ -1,52 +1,90 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-@unix_only begin
```
why did you remove this? msvcrt has `_pipe` but it has extra arguments: https://msdn.microsoft.com/en-us/library/aa298531(v=vs.60).aspx

vtjnash replied:
the eventual goal is to get this test running on windows too

tkelman replied:
not a bad goal, but will need to account for different spellings/api of posix functions - unless libuv has wrappers for all of these, which I think is one of the things we've been meaning to do upstream?

vtjnash committed:
fix the windows tests for pollfd

also works-around an issue where libuv timers could return too early on windows by several milliseconds

this provides a general-purpose socketpair implementation on windows, for use in the tests and otherwise (although uv_pipe_link is recommended for any case where you don't explicitly need a WSA2 SOCKET)

vtjnash committed:
update pollfd docs based on reviewer comments

vtjnash committed:
provide GetLastError/FormatMessage as official functions of the [Windows] Libc module (also use them to provide better errors in the tests)

these belong in Libc since the are the windows counterparts to errno/strerror which are documented and exported in Libc

vtjnash commented:
bump. this passes CI everywhere now...

JeffBezanson commented on doc/stdlib/io-network.rst:
```diff
@@ -723,22 +723,35 @@ Network I/O
-   Monitor a file for changes by polling every `interval_seconds` seconds for `seconds` seconds. A return value of true indicates
-   the file changed, a return value of false indicates a timeout.
+   Monitor a file for changes by polling every ``interval_seconds`` seconds until a change occurs or ``timeout_s`` seconds have elapsed.
+   The `timeout_s` should be a long interval; the default is 5.007 seconds.
```
Should say interval instead of timeout_s.

JeffBezanson commented on src/support/wsasocketpair.c:
```diff
@@ -0,0 +1,83 @@
+#include <io.h>
+#define AF_UNIX 1
+
+__declspec(dllexport) int wsasocketpair(int domain, int type, int protocol, SOCKET socket_vector[2]) {
```
Is this file only used for testing? Will it be useful for anything else?

vtjnash replied:
from our libuv fork, we have the equivalent of `pipe` that returns NamedPipe objects. this provides the ability to have the equivalent of `socketpair` instead. I suspect both would be useful to get pushed upstream into libuv.

ViralBShah commented:
Marking this as 0.4 - assuming this is is done and long standing, and we should get it in.

vtjnash committed:
fix poll_file variable name in doc

vtjnash commented:
bump. i'll merge this later today.

amitmurthy commented:
Awesome. :+1:

tkelman commented:
travis failure here https://gist.github.com/tkelman/96b21860b35454115373, are we back to timings that are too picky for busy low-memory VM's?

amitmurthy commented:
OTOH, in this case it does seem like it caught an actual bug. The poll call should not have returned before `intvl` seconds.

tkelman commented:
Another failure: https://travis-ci.org/JuliaLang/julia/jobs/73427729

```
exception on 5: ERROR: LoadError: test failed: 0.1 <= 0.099903232
 in expression: intvl <= t_elapsed
 in yieldto at ./task.jl:24
 in wait at ./task.jl:313
 in wait at ./task.jl:228
 in wait at ./task.jl:61
 in sync_end at ./task.jl:340
 in anonymous at no file:348
 in include_string at loading.jl:158
 in include_from_node1 at ./loading.jl:200
 in runtests at /tmp/julia/share/julia/test/testdefs.jl:173
 in anonymous at multi.jl:909
 in run_work_thunk at multi.jl:648
 in anonymous at task.jl:909
while loading /tmp/julia/share/julia/test/pollfd.jl, in expression starting on line 65
	From worker 3:	     * bitarray              in  90.72 seconds
	From worker 4:	     * mpfr                  in   4.91 seconds
	From worker 2:	     * subarray              in 353.43 seconds
ERROR: LoadError: LoadError: test failed: 0.1 <= 0.099903232
 in expression: intvl <= t_elapsed
 in anonymous at task.jl:1442
while loading /tmp/julia/share/julia/test/pollfd.jl, in expression starting on line 65
while loading /tmp/julia/share/julia/test/runtests.jl, in expression starting on line 5
	From worker 5:	     * pollfd          
```

tkelman commented:
another pollfd failure on travis https://s3.amazonaws.com/archive.travis-ci.org/jobs/75113270/log.txt

timholy commented:
Another pollfd failure on AppVeyor:
![image](https://cloud.githubusercontent.com/assets/1525481/9225657/414967ca-40d0-11e5-840d-8578228e5bea.png)
in #12560.

vtjnash commented:
i'm not really sure what that test is doing in the pollfd.jl file, since it is testing the functionality of one-shot Timers, not any of this code

@JeffBezanson @Keno is 0.2 sec sometimes not enough for a 0.01 sec timer to trigger? (added in 7f4c2de4ee7ac6403386b8d67c78826a4db36859)

tkelman commented:
same `Any[] == [1]` pollfd failure on a buildbot http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu14.04-x86/builds/2163/steps/shell_2/logs/stdio

vtjnash commented:
latest travis failure on #12463 (https://travis-ci.org/JuliaLang/julia/builds/76174785) was also due to Jeff's new test for Timers there
